{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d12bfbba-12ff-4505-b177-4f6bd88267e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import gget\n",
    "import glob\n",
    "import networkx as nx\n",
    "import os\n",
    "import time\n",
    "import seaborn as sns\n",
    "import gget\n",
    "import scanpy as sc\n",
    "import anndata as an\n",
    "from scipy.sparse import csr_matrix\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy\n",
    "import scanpy.external as sce\n",
    "\n",
    "sc.settings.verbosity = 3  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06a23093-efd0-4f94-a7dd-3ed2e31f1713",
   "metadata": {},
   "source": [
    "# Load sc data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7541c204-d7b4-487d-9543-aeb4be6c2cfb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Memory usage: current 7.93 GB, difference +7.93 GB\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 16130 × 13579\n",
       "    obs: 'cluster_str', 'dataset', 'pred_phase', 'leiden', 'n_genes', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'pct_counts_hb', 'new_cluster', 'UMAP 1', 'UMAP 2'\n",
       "    var: 'n_cells', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection'\n",
       "    uns: 'cluster_str_colors', 'dataset_colors', 'hvg', 'log1p', 'neighbors', 'new_cluster', 'new_cluster_colors', 'pca', 'pred_phase_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_scanorama', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'combat', 'counts', 'log_norm', 'magic', 'norm'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fpath = \"/scratch/indikar_root/indikar1/cstansbu/HSC/scanpy/merged_scfib.h5ad\"\n",
    "adata = sc.read_h5ad(fpath)\n",
    "adata.obs['UMAP 1'] = adata.obsm['X_umap'][:, 0]\n",
    "adata.obs['UMAP 2'] = adata.obsm['X_umap'][:, 1]\n",
    "sc.logging.print_memory_usage()\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f494beb4-9940-4104-aecb-9f0663131aec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TGTGCTGGTTGCTTAA_scFib'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs_names[-10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74e1b1c8-098a-4e4e-ba8c-643dda6c0b06",
   "metadata": {},
   "source": [
    "# Load isoquant data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c3dc9351-64a7-4fce-8cdd-44c9daeaa6dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Processing: scFib\n",
      "Loading parquet file: /scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/isoforms/isoform_data.parquet\n",
      "  - Initial shape: (75710721, 4)\n",
      "Dropping duplicate rows...\n",
      "  - Shape after deduplication: (75710721, 4)\n",
      "Creating 'barcode' column...\n",
      "Filtering rows based on 'adata.obs_names'...\n",
      "  - Final shape: (75004573, 4)\n",
      "Grouping transcripts by UMI...\n",
      "Pivoting transcripts as matrix...\n",
      "x.shape=(7748, 82648)\n",
      "Time taken for scFib: 421.52 seconds\n",
      "\n",
      "Processing: iHSC\n",
      "Loading parquet file: /scratch/indikar_root/indikar1/cstansbu/HSC/isoforms/isoform_data.parquet\n",
      "  - Initial shape: (47827803, 4)\n",
      "Dropping duplicate rows...\n",
      "  - Shape after deduplication: (47827803, 4)\n",
      "Creating 'barcode' column...\n",
      "Filtering rows based on 'adata.obs_names'...\n",
      "  - Final shape: (47634864, 4)\n",
      "Grouping transcripts by UMI...\n",
      "Pivoting transcripts as matrix...\n",
      "x.shape=(8382, 84314)\n",
      "Time taken for iHSC: 317.88 seconds\n",
      "result.shape=(16130, 99720)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>transcript_name</th>\n",
       "      <th>A1BG-201</th>\n",
       "      <th>A1BG-202</th>\n",
       "      <th>A1BG-203</th>\n",
       "      <th>A1BG-204</th>\n",
       "      <th>A2M-201</th>\n",
       "      <th>A2M-202</th>\n",
       "      <th>A2M-204</th>\n",
       "      <th>A2M-205</th>\n",
       "      <th>A2M-206</th>\n",
       "      <th>A2M-207</th>\n",
       "      <th>...</th>\n",
       "      <th>ZW10-202</th>\n",
       "      <th>ZWILCH-204</th>\n",
       "      <th>ZWILCH-206</th>\n",
       "      <th>ZWILCH-213</th>\n",
       "      <th>ZWILCH-214</th>\n",
       "      <th>ZZEF1-203</th>\n",
       "      <th>ZZEF1-205</th>\n",
       "      <th>ZZEF1-207</th>\n",
       "      <th>ZZEF1-209</th>\n",
       "      <th>ZZZ3-207</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGGGTAGCA_scFib</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGTAAGGGT_scFib</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCATTCAGGTAGG_scFib</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCATTCCAGCCCT_scFib</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCATTCGTGACCG_scFib</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 99720 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "transcript_name         A1BG-201  A1BG-202  A1BG-203  A1BG-204  A2M-201  \\\n",
       "barcode                                                                   \n",
       "AAACCAAAGGGTAGCA_scFib         0         0         0         1        0   \n",
       "AAACCAAAGTAAGGGT_scFib         0         0         0         1        0   \n",
       "AAACCATTCAGGTAGG_scFib         0         0         0         1        0   \n",
       "AAACCATTCCAGCCCT_scFib         0         0         0         1        0   \n",
       "AAACCATTCGTGACCG_scFib         0         0         0         1        1   \n",
       "\n",
       "transcript_name         A2M-202  A2M-204  A2M-205  A2M-206  A2M-207  ...  \\\n",
       "barcode                                                              ...   \n",
       "AAACCAAAGGGTAGCA_scFib        0        0        0        0        0  ...   \n",
       "AAACCAAAGTAAGGGT_scFib        0        0        0        0        0  ...   \n",
       "AAACCATTCAGGTAGG_scFib        0        0        0        0        0  ...   \n",
       "AAACCATTCCAGCCCT_scFib        0        0        0        0        0  ...   \n",
       "AAACCATTCGTGACCG_scFib        0        0        0        0        0  ...   \n",
       "\n",
       "transcript_name         ZW10-202  ZWILCH-204  ZWILCH-206  ZWILCH-213  \\\n",
       "barcode                                                                \n",
       "AAACCAAAGGGTAGCA_scFib       0.0         0.0         0.0         0.0   \n",
       "AAACCAAAGTAAGGGT_scFib       0.0         0.0         0.0         0.0   \n",
       "AAACCATTCAGGTAGG_scFib       0.0         0.0         0.0         0.0   \n",
       "AAACCATTCCAGCCCT_scFib       0.0         0.0         0.0         0.0   \n",
       "AAACCATTCGTGACCG_scFib       0.0         0.0         0.0         0.0   \n",
       "\n",
       "transcript_name         ZWILCH-214  ZZEF1-203  ZZEF1-205  ZZEF1-207  \\\n",
       "barcode                                                               \n",
       "AAACCAAAGGGTAGCA_scFib         0.0        0.0        0.0        0.0   \n",
       "AAACCAAAGTAAGGGT_scFib         0.0        0.0        0.0        0.0   \n",
       "AAACCATTCAGGTAGG_scFib         0.0        0.0        0.0        0.0   \n",
       "AAACCATTCCAGCCCT_scFib         0.0        0.0        0.0        0.0   \n",
       "AAACCATTCGTGACCG_scFib         0.0        0.0        0.0        0.0   \n",
       "\n",
       "transcript_name         ZZEF1-209  ZZZ3-207  \n",
       "barcode                                      \n",
       "AAACCAAAGGGTAGCA_scFib        0.0       0.0  \n",
       "AAACCAAAGTAAGGGT_scFib        0.0       0.0  \n",
       "AAACCATTCAGGTAGG_scFib        0.0       0.0  \n",
       "AAACCATTCCAGCCCT_scFib        0.0       0.0  \n",
       "AAACCATTCGTGACCG_scFib        0.0       0.0  \n",
       "\n",
       "[5 rows x 99720 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fpaths = {\n",
    "    'scFib' : '/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/isoforms/isoform_data.parquet',\n",
    "    'iHSC' : '/scratch/indikar_root/indikar1/cstansbu/HSC/isoforms/isoform_data.parquet',\n",
    "}\n",
    "\n",
    "columns = [\n",
    "    'cell_id',\n",
    "    'UMI',\n",
    "    'gene_name',\n",
    "    'transcript_name',\n",
    "]\n",
    "\n",
    "\n",
    "result = []\n",
    "\n",
    "for k, v in fpaths.items():\n",
    "    start_time = time.time()\n",
    "    print(f\"\\nProcessing: {k}\")\n",
    "    print(f\"Loading parquet file: {v}\")\n",
    "    df = pd.read_parquet(v, columns=columns)\n",
    "    print(f\"  - Initial shape: {df.shape}\")\n",
    "\n",
    "    print(\"Dropping duplicate rows...\")\n",
    "    df = df.drop_duplicates()\n",
    "    print(f\"  - Shape after deduplication: {df.shape}\")\n",
    "\n",
    "    print(\"Creating 'barcode' column...\")\n",
    "    df['barcode'] = df['cell_id'] + f\"_{k}\"\n",
    "    del df['cell_id']\n",
    "\n",
    "    print(\"Filtering rows based on 'adata.obs_names'...\") \n",
    "    df = df[df['barcode'].isin(adata.obs_names)]\n",
    "    print(f\"  - Final shape: {df.shape}\")\n",
    "    \n",
    "    \n",
    "    print(\"Grouping transcripts by UMI...\") \n",
    "    x = df.groupby(['barcode', 'transcript_name'])['UMI'].nunique().reset_index()\n",
    "    \n",
    "    print(\"Pivoting transcripts as matrix...\") \n",
    "    x = pd.pivot_table(\n",
    "        x, \n",
    "        index='barcode', \n",
    "        columns='transcript_name',\n",
    "        values='UMI',\n",
    "        fill_value=0.0,\n",
    "    )\n",
    "\n",
    "    print(f\"{x.shape=}\")\n",
    "    result.append(x)\n",
    "\n",
    "    end_time = time.time()\n",
    "    print(f\"Time taken for {k}: {end_time - start_time:.2f} seconds\")\n",
    "    \n",
    "    \n",
    "result = pd.concat(result)\n",
    "result = result.fillna(0.0)\n",
    "print(f\"{result.shape=}\")\n",
    "result.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6988a267-2716-4aa8-91b7-c0eaf629213b",
   "metadata": {},
   "source": [
    "# BUILD BDATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "19b2a4e9-5997-44aa-b341-94140ee93a08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 16130 × 99720\n",
       "    layers: 'raw_counts'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdata = an.AnnData(csr_matrix(result.to_numpy()))\n",
    "bdata.obs_names = result.index\n",
    "bdata.var_names = result.columns\n",
    "\n",
    "bdata.layers['raw_counts'] = csr_matrix(bdata.X.copy())\n",
    "\n",
    "bdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0c68bbf9-3c0e-420a-acf4-3d04d2c17ba4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_count</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_count</th>\n",
       "      <th>transcript_percent</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG-201</th>\n",
       "      <td>4.0</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>11582.0</td>\n",
       "      <td>0.000345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-202</th>\n",
       "      <td>9.0</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>11582.0</td>\n",
       "      <td>0.000777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-203</th>\n",
       "      <td>68.0</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>11582.0</td>\n",
       "      <td>0.005871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-204</th>\n",
       "      <td>11501.0</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>11582.0</td>\n",
       "      <td>0.993006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-201</th>\n",
       "      <td>1368.0</td>\n",
       "      <td>A2M</td>\n",
       "      <td>1514.0</td>\n",
       "      <td>0.903567</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 transcript_count gene_name  gene_count  transcript_percent\n",
       "transcript_name                                                            \n",
       "A1BG-201                      4.0      A1BG     11582.0            0.000345\n",
       "A1BG-202                      9.0      A1BG     11582.0            0.000777\n",
       "A1BG-203                     68.0      A1BG     11582.0            0.005871\n",
       "A1BG-204                  11501.0      A1BG     11582.0            0.993006\n",
       "A2M-201                    1368.0       A2M      1514.0            0.903567"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var = result.sum(axis=0).reset_index()\n",
    "var.columns = ['transcript_name', 'transcript_count']\n",
    "var['gene_name'] = var['transcript_name'].apply(lambda x: x.split(\"-\")[0])\n",
    "var['gene_count'] = var.groupby('gene_name')['transcript_count'].transform('sum')\n",
    "var['transcript_percent'] = var['transcript_count'] / var['gene_count']\n",
    "var = var.set_index('transcript_name')\n",
    "var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bc63f10a-836e-4270-b028-675874a32b2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_count</th>\n",
       "      <th>cluster_str</th>\n",
       "      <th>dataset</th>\n",
       "      <th>pred_phase</th>\n",
       "      <th>leiden</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>pct_counts_in_top_50_genes</th>\n",
       "      <th>pct_counts_in_top_100_genes</th>\n",
       "      <th>...</th>\n",
       "      <th>pct_counts_in_top_500_genes</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "      <th>total_counts_ribo</th>\n",
       "      <th>pct_counts_ribo</th>\n",
       "      <th>total_counts_hb</th>\n",
       "      <th>pct_counts_hb</th>\n",
       "      <th>new_cluster</th>\n",
       "      <th>UMAP 1</th>\n",
       "      <th>UMAP 2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGGGTAGCA_scFib</th>\n",
       "      <td>14275.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>scFib</td>\n",
       "      <td>G2M</td>\n",
       "      <td>0</td>\n",
       "      <td>3700</td>\n",
       "      <td>3700</td>\n",
       "      <td>16000</td>\n",
       "      <td>29.887500</td>\n",
       "      <td>40.806250</td>\n",
       "      <td>...</td>\n",
       "      <td>65.706250</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>47</td>\n",
       "      <td>0.293750</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0.286714</td>\n",
       "      <td>14.078166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGTAAGGGT_scFib</th>\n",
       "      <td>7402.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>scFib</td>\n",
       "      <td>G1</td>\n",
       "      <td>1</td>\n",
       "      <td>2278</td>\n",
       "      <td>2278</td>\n",
       "      <td>7862</td>\n",
       "      <td>34.991096</td>\n",
       "      <td>46.731112</td>\n",
       "      <td>...</td>\n",
       "      <td>71.712033</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14</td>\n",
       "      <td>0.178072</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>3.464725</td>\n",
       "      <td>5.875223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCATTCAGGTAGG_scFib</th>\n",
       "      <td>13329.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>scFib</td>\n",
       "      <td>G2M</td>\n",
       "      <td>0</td>\n",
       "      <td>3713</td>\n",
       "      <td>3713</td>\n",
       "      <td>15607</td>\n",
       "      <td>28.384699</td>\n",
       "      <td>38.758250</td>\n",
       "      <td>...</td>\n",
       "      <td>64.644070</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49</td>\n",
       "      <td>0.313962</td>\n",
       "      <td>1</td>\n",
       "      <td>0.006407</td>\n",
       "      <td>1</td>\n",
       "      <td>0.638370</td>\n",
       "      <td>12.075577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCATTCCAGCCCT_scFib</th>\n",
       "      <td>10586.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>scFib</td>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>2446</td>\n",
       "      <td>2446</td>\n",
       "      <td>10504</td>\n",
       "      <td>38.442498</td>\n",
       "      <td>51.266184</td>\n",
       "      <td>...</td>\n",
       "      <td>75.361767</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16</td>\n",
       "      <td>0.152323</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>3.479215</td>\n",
       "      <td>7.809453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCATTCGTGACCG_scFib</th>\n",
       "      <td>7787.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>scFib</td>\n",
       "      <td>G1</td>\n",
       "      <td>1</td>\n",
       "      <td>2580</td>\n",
       "      <td>2580</td>\n",
       "      <td>8774</td>\n",
       "      <td>29.735582</td>\n",
       "      <td>40.733987</td>\n",
       "      <td>...</td>\n",
       "      <td>67.996353</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>28</td>\n",
       "      <td>0.319125</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>3.956977</td>\n",
       "      <td>6.376747</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        transcript_count cluster_str dataset pred_phase  \\\n",
       "barcode                                                                   \n",
       "AAACCAAAGGGTAGCA_scFib           14275.0         NaN   scFib        G2M   \n",
       "AAACCAAAGTAAGGGT_scFib            7402.0         NaN   scFib         G1   \n",
       "AAACCATTCAGGTAGG_scFib           13329.0         NaN   scFib        G2M   \n",
       "AAACCATTCCAGCCCT_scFib           10586.0         NaN   scFib          S   \n",
       "AAACCATTCGTGACCG_scFib            7787.0         NaN   scFib         G1   \n",
       "\n",
       "                       leiden  n_genes  n_genes_by_counts  total_counts  \\\n",
       "barcode                                                                   \n",
       "AAACCAAAGGGTAGCA_scFib      0     3700               3700         16000   \n",
       "AAACCAAAGTAAGGGT_scFib      1     2278               2278          7862   \n",
       "AAACCATTCAGGTAGG_scFib      0     3713               3713         15607   \n",
       "AAACCATTCCAGCCCT_scFib      1     2446               2446         10504   \n",
       "AAACCATTCGTGACCG_scFib      1     2580               2580          8774   \n",
       "\n",
       "                        pct_counts_in_top_50_genes  \\\n",
       "barcode                                              \n",
       "AAACCAAAGGGTAGCA_scFib                   29.887500   \n",
       "AAACCAAAGTAAGGGT_scFib                   34.991096   \n",
       "AAACCATTCAGGTAGG_scFib                   28.384699   \n",
       "AAACCATTCCAGCCCT_scFib                   38.442498   \n",
       "AAACCATTCGTGACCG_scFib                   29.735582   \n",
       "\n",
       "                        pct_counts_in_top_100_genes  ...  \\\n",
       "barcode                                              ...   \n",
       "AAACCAAAGGGTAGCA_scFib                    40.806250  ...   \n",
       "AAACCAAAGTAAGGGT_scFib                    46.731112  ...   \n",
       "AAACCATTCAGGTAGG_scFib                    38.758250  ...   \n",
       "AAACCATTCCAGCCCT_scFib                    51.266184  ...   \n",
       "AAACCATTCGTGACCG_scFib                    40.733987  ...   \n",
       "\n",
       "                        pct_counts_in_top_500_genes  total_counts_mt  \\\n",
       "barcode                                                                \n",
       "AAACCAAAGGGTAGCA_scFib                    65.706250                0   \n",
       "AAACCAAAGTAAGGGT_scFib                    71.712033                0   \n",
       "AAACCATTCAGGTAGG_scFib                    64.644070                0   \n",
       "AAACCATTCCAGCCCT_scFib                    75.361767                0   \n",
       "AAACCATTCGTGACCG_scFib                    67.996353                0   \n",
       "\n",
       "                        pct_counts_mt  total_counts_ribo  pct_counts_ribo  \\\n",
       "barcode                                                                     \n",
       "AAACCAAAGGGTAGCA_scFib            0.0                 47         0.293750   \n",
       "AAACCAAAGTAAGGGT_scFib            0.0                 14         0.178072   \n",
       "AAACCATTCAGGTAGG_scFib            0.0                 49         0.313962   \n",
       "AAACCATTCCAGCCCT_scFib            0.0                 16         0.152323   \n",
       "AAACCATTCGTGACCG_scFib            0.0                 28         0.319125   \n",
       "\n",
       "                        total_counts_hb  pct_counts_hb  new_cluster    UMAP 1  \\\n",
       "barcode                                                                         \n",
       "AAACCAAAGGGTAGCA_scFib                0       0.000000            1  0.286714   \n",
       "AAACCAAAGTAAGGGT_scFib                0       0.000000            2  3.464725   \n",
       "AAACCATTCAGGTAGG_scFib                1       0.006407            1  0.638370   \n",
       "AAACCATTCCAGCCCT_scFib                0       0.000000            2  3.479215   \n",
       "AAACCATTCGTGACCG_scFib                0       0.000000            2  3.956977   \n",
       "\n",
       "                           UMAP 2  \n",
       "barcode                            \n",
       "AAACCAAAGGGTAGCA_scFib  14.078166  \n",
       "AAACCAAAGTAAGGGT_scFib   5.875223  \n",
       "AAACCATTCAGGTAGG_scFib  12.075577  \n",
       "AAACCATTCCAGCCCT_scFib   7.809453  \n",
       "AAACCATTCGTGACCG_scFib   6.376747  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "obs = result.sum(axis=1).reset_index()\n",
    "obs.columns = ['barcode', 'transcript_count']\n",
    "obs = obs.set_index('barcode')\n",
    "\n",
    "obs = pd.merge(\n",
    "    obs, \n",
    "    adata.obs.copy(),\n",
    "    how='left',\n",
    "    right_index=True,\n",
    "    left_index=True,\n",
    ")\n",
    "\n",
    "obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8d011249-6b2b-4872-94a7-7c33a3c24e6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 16130 × 99720\n",
       "    obs: 'transcript_count', 'cluster_str', 'dataset', 'pred_phase', 'leiden', 'n_genes', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'pct_counts_hb', 'new_cluster', 'UMAP 1', 'UMAP 2'\n",
       "    var: 'transcript_count', 'gene_name', 'gene_count', 'transcript_percent'\n",
       "    layers: 'raw_counts'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdata.var = var\n",
    "bdata.obs = obs\n",
    "\n",
    "bdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9781d465-37cb-47e8-ad97-9a6d8f1a5797",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n"
     ]
    }
   ],
   "source": [
    "# Normalizing to median total counts\n",
    "sc.pp.normalize_total(bdata, target_sum=1e6)\n",
    "bdata.layers[\"norm\"] = csr_matrix(bdata.X.astype('float32').copy())\n",
    "sc.pp.log1p(bdata)\n",
    "bdata.layers[\"log_norm\"] = csr_matrix(bdata.X.astype('float32').copy())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "001319b0-cd8c-4bbd-8c20-088e00437136",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "extracting highly variable genes\n",
      "    finished (0:00:02)\n",
      "--> added\n",
      "    'highly_variable', boolean vector (adata.var)\n",
      "    'means', float vector (adata.var)\n",
      "    'dispersions', float vector (adata.var)\n",
      "    'dispersions_norm', float vector (adata.var)\n",
      "computing PCA\n",
      "    with n_comps=50\n",
      "    finished (0:00:05)\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 50\n",
      "    finished: added to `.uns['neighbors']`\n",
      "    `.obsp['distances']`, distances for each pair of neighbors\n",
      "    `.obsp['connectivities']`, weighted adjacency matrix (0:00:28)\n",
      "computing UMAP\n",
      "    finished: added\n",
      "    'X_umap', UMAP coordinates (adata.obsm) (0:00:16)\n"
     ]
    }
   ],
   "source": [
    "sc.pp.highly_variable_genes(\n",
    "    bdata,\n",
    ")\n",
    "\n",
    "sc.tl.pca(\n",
    "    bdata,\n",
    "    mask_var='highly_variable',\n",
    ")\n",
    "\n",
    "sc.pp.neighbors(bdata,)\n",
    "\n",
    "sc.tl.umap(\n",
    "    bdata,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "46dc748b-fb34-43dd-aaf8-20b7c6f29076",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'gene_name' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Standardizing Data across genes.\n",
      "\n",
      "Found 2 batches\n",
      "\n",
      "Found 2 numerical variables:\n",
      "\ttotal_counts, n_genes\n",
      "\n",
      "Fitting L/S model and finding priors\n",
      "\n",
      "Finding parametric adjustments\n",
      "\n",
      "Adjusting data\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 16130 × 99720\n",
       "    obs: 'transcript_count', 'cluster_str', 'dataset', 'pred_phase', 'leiden', 'n_genes', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'pct_counts_hb', 'new_cluster', 'UMAP 1', 'UMAP 2'\n",
       "    var: 'transcript_count', 'gene_name', 'gene_count', 'transcript_percent', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'log1p', 'hvg', 'pca', 'neighbors', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'raw_counts', 'norm', 'log_norm', 'combat'\n",
       "    obsp: 'distances', 'connectivities'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combat\n",
    "sc.pp.combat(\n",
    "    bdata, \n",
    "    key='dataset',\n",
    "    covariates=['total_counts', 'n_genes'],\n",
    ")\n",
    "\n",
    "bdata.X = csr_matrix(bdata.X.astype('float32'))\n",
    "bdata.layers[\"combat\"] = bdata.X.copy()\n",
    "\n",
    "# set current layer\n",
    "bdata.X = bdata.layers[\"raw_counts\"]\n",
    "bdata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb531bc5-6e5e-4af5-9715-59be7826da20",
   "metadata": {},
   "source": [
    "# MAGIC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "798e1da8-54d7-40ac-a0b1-58175c5787c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing MAGIC\n",
      "    finished (0:01:59)\n"
     ]
    }
   ],
   "source": [
    "sce.pp.magic(\n",
    "    bdata,\n",
    "    name_list='all_genes',\n",
    "    solver='approximate',\n",
    "    copy=False,\n",
    ")\n",
    "\n",
    "bdata.layers[\"magic\"] = csr_matrix(bdata.X.astype('float32').copy())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "98ddb171-a8d3-403c-ad98-f7965817aecb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 16130 × 99720\n",
       "    obs: 'transcript_count', 'cluster_str', 'dataset', 'pred_phase', 'leiden', 'n_genes', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'pct_counts_hb', 'new_cluster', 'UMAP 1', 'UMAP 2'\n",
       "    var: 'transcript_count', 'gene_name', 'gene_count', 'transcript_percent', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'log1p', 'hvg', 'pca', 'neighbors', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'raw_counts', 'norm', 'log_norm', 'combat', 'magic'\n",
       "    obsp: 'distances', 'connectivities'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "44097ba8-ba00-4bab-9409-05a9b3fa91bc",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "'break' outside loop (668683560.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[14], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    break\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m 'break' outside loop\n"
     ]
    }
   ],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "62f64727-dc88-4988-aa77-34219cfd4ade",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 16130 × 99720\n",
       "    obs: 'transcript_count', 'cluster_str', 'dataset', 'pred_phase', 'leiden', 'n_genes', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'pct_counts_hb', 'new_cluster', 'UMAP 1', 'UMAP 2'\n",
       "    var: 'transcript_count', 'gene_name', 'gene_count', 'transcript_percent', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'log1p', 'hvg', 'pca', 'neighbors', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'raw_counts', 'norm', 'log_norm', 'combat', 'magic'\n",
       "    obsp: 'distances', 'connectivities'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outpath = \"/scratch/indikar_root/indikar1/cstansbu/HSC/scanpy/merged_isoforms.h5ad\"\n",
    "bdata.write(outpath)\n",
    "\n",
    "bdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5cef93a-ee19-480d-ae00-8922c9499187",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1ea852f-d76d-4805-adff-7ed18abee227",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02eceac9-0a39-4a4a-9c89-bdbf9e202489",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a562db59-dafa-46fc-9ac3-8bc4b99b622e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ecba1f7-5dd3-45ee-9e10-244e73663dd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d320cec-64b0-4a5f-b971-464d9a947140",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = df.groupby(['barcode', 'transcript_name'])['UMI'].nunique().reset_index()\n",
    "\n",
    "x = pd.pivot_table(\n",
    "    x, \n",
    "    index='barcode', \n",
    "    columns='transcript_name',\n",
    "    values='UMI',\n",
    "    fill_value=0.0,\n",
    ")\n",
    "\n",
    "print(f\"{x.shape=}\")\n",
    "x.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd315211-9246-4fca-a7ea-dd53ae4ece7e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4a3db74-ad27-4a72-8ba8-f7a6bf3d8715",
   "metadata": {},
   "outputs": [],
   "source": [
    "list(df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a07ab73e-39af-4798-bfa8-1a1ff05941f1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31d416ae-b6be-49a4-a4fc-282dbae5cbce",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e3ea51c-5b7e-44e8-89de-96b310545599",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d5dad75-221c-4f7c-ae32-4ba946036367",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c0e77af-152d-492d-bf15-f9ad846925b4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f0d0df4-a096-49c2-b068-eadfe22f6976",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cefa3fe0-1a54-42e5-b4dd-6ffef1a42b32",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3eb87290-4e47-49ed-aad3-f8e69dbdcdcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "fpath = \"/scratch/indikar_root/indikar1/cstansbu/HSC/isoforms/isoform_data.parquet\"\n",
    "\n",
    "df = pd.read_parquet(fpath)\n",
    "df = df[df['cell_id'].isin(adata.obs_names)] # filter out undetected CBs\n",
    "\n",
    "# add nulti-map flag\n",
    "df['n_isoforms'] = df.groupby('gene_name')['transcript_name'].transform('nunique')\n",
    "\n",
    "print(f\"{df.shape=}\")\n",
    "df = pd.merge(\n",
    "    df, \n",
    "    adata.obs,\n",
    "    how='left',\n",
    "    left_on='cell_id',\n",
    "    right_index=True,\n",
    ")\n",
    "\n",
    "outpath = \"/nfs/turbo/umms-indikar/shared/projects/HSC/data/isoforms/isoquant.parquet\"\n",
    "df.to_parquet(outpath, index=False)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67bc0562-f461-4f30-ba4f-5e4c3848668c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "scanpy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
