{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8516067d-bc9c-492f-810f-f570f0093b02",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1da85bb0-7483-4421-af62-54e04f775bea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Run7x0315.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Run4x0314.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Run6x0314.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Gridr4.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Run4x0422.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/P2r6.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/P2r7.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Run1x0312.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Run8x0426.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Run2x0313.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Gridr3.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Run2x0416.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Run1x0415.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Run5x0314.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Run3x0314.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Run3x0418.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/P2r5.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Gridr1.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Run7x0426.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Run6x0426.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Run9x0503.tagged.csv',\n",
       " '/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/Gridr2.tagged.csv']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dpath = \"/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/\"\n",
    "file_list = glob.glob(f\"{dpath}*.tagged.csv\")\n",
    "file_list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f9f7ebea-2d26-4908-b0be-941ac52397a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "def create_read_map(read_map_path: str) -> dict[str, str]:\n",
    "    \"\"\"Creates a dictionary mapping query names to gene names from a CSV file.\n",
    "\n",
    "    Args:\n",
    "        read_map_path (str): The file path to the CSV containing the mapping.\n",
    "\n",
    "    Returns:\n",
    "        dict: A dictionary where keys are query names and values are gene names.\n",
    "    \"\"\"\n",
    "    \n",
    "    read_map_df = pd.read_csv(read_map_path)\n",
    "    return read_map_df.set_index('query_name')['gene_name'].to_dict()\n",
    "\n",
    "\n",
    "read_map_path = \"/scratch/indikar_root/indikar1/cstansbu/HSC/v5_tagged/read_map.csv\"\n",
    "read_map = create_read_map(read_map_path)\n",
    "\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fb40ef5e-ec00-4f38-afdc-30b3387e9f31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HT-FOS</th>\n",
       "      <th>HT-GATA2</th>\n",
       "      <th>HT-REL</th>\n",
       "      <th>HT-STAT5A</th>\n",
       "      <th>V5-FOS</th>\n",
       "      <th>V5-GATA2</th>\n",
       "      <th>V5-REL</th>\n",
       "      <th>V5-STAT5A</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACGCTTCACTCCGT</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACAAAGGTACTAGCT</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACCATGAGCTGGCCT</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACGGGACATCAGCAT</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACTTCTAGCTACGTT</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  HT-FOS  HT-GATA2  HT-REL  HT-STAT5A  V5-FOS  V5-GATA2  \\\n",
       "barcode                                                                   \n",
       "AAACGCTTCACTCCGT       1         0       0          0       1         0   \n",
       "AACAAAGGTACTAGCT       1         0       0          0       1         0   \n",
       "AACCATGAGCTGGCCT       0         0       0          0       1         0   \n",
       "AACGGGACATCAGCAT       0         0       0          0       0         1   \n",
       "AACTTCTAGCTACGTT       0         1       0          0       0         1   \n",
       "\n",
       "                  V5-REL  V5-STAT5A  \n",
       "barcode                              \n",
       "AAACGCTTCACTCCGT       0          0  \n",
       "AACAAAGGTACTAGCT       0          0  \n",
       "AACCATGAGCTGGCCT       0          0  \n",
       "AACGGGACATCAGCAT       0          0  \n",
       "AACTTCTAGCTACGTT       0          0  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def parse_seqID(input_string):\n",
    "    \"\"\"\n",
    "    Parse an input string containing barcode, UMI, and read name.\n",
    "\n",
    "    Args:\n",
    "        input_string (str): A string containing barcode, UMI, and read name separated by '#' and '_'.\n",
    "\n",
    "    Returns:\n",
    "        tuple: A tuple containing barcode, UMI, and read name.\n",
    "    \"\"\"\n",
    "    parts = input_string.split('#')\n",
    "    barcode_umi = parts[0]\n",
    "    barcode, umi = barcode_umi.split('_')\n",
    "    read_name = parts[1].split('_')[0]\n",
    "    return barcode, umi, read_name\n",
    "\n",
    "\n",
    "def process_read_metadata(tmp: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"Processes read metadata to identify V5 and HT patterns in gene groups.\n",
    "\n",
    "    Args:\n",
    "        tmp (pd.DataFrame): Input DataFrame containing read metadata, \n",
    "                            including columns 'read_name', 'barcode', 'umi', 'strand', 'gene_name', and 'patternName'.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: A DataFrame summarizing the presence of V5 and HT patterns for each unique combination of \n",
    "                      read name, barcode, UMI, strand, and gene name.\n",
    "\n",
    "    \"\"\"\n",
    "    res = []\n",
    "\n",
    "    for read_meta, group in tmp.groupby(['read_name', 'barcode', 'umi', 'strand', 'gene_name']):\n",
    "        read_name, barcode, umi, strand, gene_name = read_meta\n",
    "\n",
    "        has_V5 = 'V5' in group['patternName'].str.strip().values\n",
    "        has_HT = 'HT' in group['patternName'].str.strip().values\n",
    "\n",
    "        row = {\n",
    "            'barcode': barcode,\n",
    "            'umi': umi,\n",
    "            'read_name': read_name,\n",
    "            'gene_name': gene_name,\n",
    "            'has_V5': has_V5,\n",
    "            'has_HT': has_HT,\n",
    "        }\n",
    "        res.append(row)\n",
    "\n",
    "    return pd.DataFrame(res)\n",
    "\n",
    "\n",
    "def count_has_vh_by_unique_umi(df):\n",
    "    \"\"\"Counts the number of has_V5 and has_HT columns by unique UMIs per barcode per gene_name.\n",
    "\n",
    "    Args:\n",
    "    df: A pandas DataFrame with columns 'barcode', 'umi', 'read_name', 'gene_name',\n",
    "        'has_V5', and 'has_HT'.\n",
    "\n",
    "    Returns:\n",
    "    A pandas DataFrame with columns 'barcode', 'gene_name', 'num_has_V5', 'num_has_HT',\n",
    "        where 'num_has_V5' is the count of True values in the 'has_V5' column for\n",
    "        each unique combination of 'barcode' and 'gene_name', and 'num_has_HT' is\n",
    "        similarly defined for the 'has_HT' column.\n",
    "    \"\"\"\n",
    "\n",
    "    return (df.groupby(['barcode', 'gene_name'])\n",
    "          .agg(num_V5=('has_V5', sum), num_HT=('has_HT', sum))).reset_index()\n",
    "\n",
    "\n",
    "def process_barcode_search(file_list, read_map):\n",
    "    \"\"\"\n",
    "    Processes sequencing files to extract barcode information, map reads to genes, and\n",
    "    summarize the presence of V5 and HT tags.\n",
    "\n",
    "    This function iterates through a list of sequencing files, parses the sequence IDs to\n",
    "    extract barcode, UMI, and read name information. It then maps reads to their corresponding\n",
    "    gene names and processes read metadata to identify V5 and HT tags. Finally, it aggregates\n",
    "    the results into a DataFrame.\n",
    "\n",
    "    Args:\n",
    "        file_list (list): A list of file paths to sequencing data files.\n",
    "        read_map (dict): A dictionary mapping read IDs to their corresponding gene names.\n",
    "\n",
    "    Returns:\n",
    "        pandas.DataFrame: A DataFrame containing the processed data, including columns for\n",
    "            barcode, gene name, and counts of V5 and HT tags.\n",
    "\n",
    "    Example:\n",
    "        result_df = process_barcode_search([\"file1.txt\", \"file2.txt\"], read_map)\n",
    "    \"\"\"\n",
    "\n",
    "    df = []\n",
    "\n",
    "    for fpath in file_list:\n",
    "        file_id = os.path.basename(fpath).split(\".\")[0]\n",
    "\n",
    "        # load data \n",
    "        tmp = pd.read_csv(fpath, sep=\"\\t\")\n",
    "        if tmp.shape[0] == 0:\n",
    "            continue\n",
    "\n",
    "        tmp[['barcode', 'umi', 'read_name']] = tmp['seqID'].apply(parse_seqID).apply(pd.Series)\n",
    "\n",
    "        # map the reads to the genes that they aligned to\n",
    "        tmp['gene_name'] = tmp['seqID'].map(read_map)\n",
    "\n",
    "        # create boolean flags for each detected V5 tag\n",
    "        tmp = process_read_metadata(tmp)\n",
    "        tmp = count_has_vh_by_unique_umi(tmp)\n",
    "        tmp['file_id'] = file_id\n",
    "        df.append(tmp)\n",
    "\n",
    "    df = pd.concat(df)\n",
    "    return df\n",
    "\n",
    "\n",
    "def calculate_sum_by_gene_barcode(df):\n",
    "    \"\"\"\n",
    "    Calculates the sum of 'num_V5' and 'num_HT' for each gene by barcode,\n",
    "    aggregating over all 'file_id'.\n",
    "\n",
    "    Args:\n",
    "        df: A pandas DataFrame with columns 'barcode', 'gene_name', 'num_V5', 'num_HT', and 'file_id'.\n",
    "\n",
    "    Returns:\n",
    "        A pandas DataFrame with columns 'barcode', 'gene_name', 'num_V5', and 'num_HT', \n",
    "        where 'num_V5' and 'num_HT' are the summed values.\n",
    "    \"\"\"\n",
    "\n",
    "    # Group by 'barcode' and 'gene_name', then sum the relevant columns\n",
    "    result_df = df.groupby(['barcode', 'gene_name'])[['num_V5', 'num_HT']].sum().reset_index()\n",
    "    \n",
    "    result_df = pd.pivot_table(result_df, index='barcode',\n",
    "                               columns='gene_name',\n",
    "                               fill_value=0)\n",
    "\n",
    "    result_df.columns = ['-'.join(col).strip().replace(\"num_\", \"\") for col in result_df.columns.values]\n",
    "    return result_df\n",
    "\n",
    "\n",
    "\n",
    "df = process_barcode_search(file_list, read_map)\n",
    "df = calculate_sum_by_gene_barcode(df)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bde2433-25a9-4cff-b302-b500f20030eb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
