{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c35b7e6c-a4a1-4928-9b27-a7c24065d1c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import anndata as ad\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5392c223-3667-487d-9ba3-805538b74850",
   "metadata": {},
   "source": [
    "# PangloaDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a6a00e0e-012d-425f-b4d4-e7550f435f5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>ubiquitousness_index</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>organ</th>\n",
       "      <th>sensitivity_human</th>\n",
       "      <th>specificity_human</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>531</th>\n",
       "      <td>CD38</td>\n",
       "      <td>B cells memory</td>\n",
       "      <td>0.016</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>Immune system</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.014098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>532</th>\n",
       "      <td>CD80</td>\n",
       "      <td>B cells memory</td>\n",
       "      <td>0.000</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>Immune system</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>533</th>\n",
       "      <td>CD84</td>\n",
       "      <td>B cells memory</td>\n",
       "      <td>0.023</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>Immune system</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.009399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534</th>\n",
       "      <td>CD86</td>\n",
       "      <td>B cells memory</td>\n",
       "      <td>0.026</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>Immune system</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.012218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>535</th>\n",
       "      <td>NT5E</td>\n",
       "      <td>B cells memory</td>\n",
       "      <td>0.005</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>Immune system</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007206</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_name       cell_type  ubiquitousness_index            gene_type  \\\n",
       "531      CD38  B cells memory                 0.016  protein-coding gene   \n",
       "532      CD80  B cells memory                 0.000  protein-coding gene   \n",
       "533      CD84  B cells memory                 0.023  protein-coding gene   \n",
       "534      CD86  B cells memory                 0.026  protein-coding gene   \n",
       "535      NT5E  B cells memory                 0.005  protein-coding gene   \n",
       "\n",
       "             organ  sensitivity_human  specificity_human  \n",
       "531  Immune system                0.0           0.014098  \n",
       "532  Immune system                0.0           0.000000  \n",
       "533  Immune system                0.0           0.009399  \n",
       "534  Immune system                0.0           0.012218  \n",
       "535  Immune system                0.0           0.007206  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pangloa_path = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/panglaodb/pandb.tsv.gz\"\n",
    "pdf = pd.read_csv(pangloa_path, sep=\"\\t\")\n",
    "\n",
    "pdf = pdf[pdf['species'].str.contains('Hs')]\n",
    "\n",
    "cell_types = [\n",
    "    'Fibroblasts',\n",
    "    'Hematopoietic stem cells',\n",
    "    \"Erythroblasts\",\n",
    "    \"Megakaryocytes\",\n",
    "    \"Myeloid-derived suppressor cells\",\n",
    "    \"B cells memory\",\n",
    "    \"B cells naive\",\n",
    "    \"Plasma cells\",\n",
    "    \"Monocytes\",\n",
    "    \"Macrophages\",\n",
    "    \"Dendritic cells\",\n",
    "    \"Plasmacytoid dendritic cells\",\n",
    "    \"Neutrophils\",\n",
    "    \"Eosinophils\",\n",
    "    \"Basophils\",\n",
    "    \"Reticulocytes\",\n",
    "    \"Stromal cells\",\n",
    "    \"Osteoblasts\",\n",
    "    \"Endothelial cells\",\n",
    "]\n",
    "\n",
    "columns = [\n",
    "    'official gene symbol',\n",
    "    'cell type',\n",
    "    'ubiquitousness index',\n",
    "    'gene type',\n",
    "    'organ',\n",
    "    'sensitivity_human',\n",
    "    'specificity_human',\n",
    "]\n",
    "\n",
    "pdf = pdf[pdf['cell type'].isin(cell_types)]\n",
    "pdf = pdf[columns]\n",
    "\n",
    "pdf.columns = [\n",
    "    'gene_name',\n",
    "    'cell_type',\n",
    "    'ubiquitousness_index',\n",
    "    'gene_type',\n",
    "    'organ',\n",
    "    'sensitivity_human',\n",
    "    'specificity_human'\n",
    "]\n",
    "\n",
    "\n",
    "output_path = \"../config/gene_annotations/panglaodb.csv\"\n",
    "pdf.to_csv(output_path, index=False)\n",
    "pdf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9eb8fa49-c2e5-49ec-954e-3ca5cd9c80fb",
   "metadata": {},
   "source": [
    "# SCENIC transcription factors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "92e1fae5-a768-4a2f-b350-77498ed1e2da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len(genes)=1892\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZNF354C</td>\n",
       "      <td>transcription_factor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KLF12</td>\n",
       "      <td>transcription_factor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZNF143</td>\n",
       "      <td>transcription_factor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZIC2</td>\n",
       "      <td>transcription_factor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZNF274</td>\n",
       "      <td>transcription_factor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name             gene_type\n",
       "0   ZNF354C  transcription_factor\n",
       "1     KLF12  transcription_factor\n",
       "2    ZNF143  transcription_factor\n",
       "3      ZIC2  transcription_factor\n",
       "4    ZNF274  transcription_factor"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fpath = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/scenic/tf_lists/allTFs_hg38.txt\"\n",
    "\n",
    "genes = [x.strip() for x in open(fpath)]\n",
    "print(f\"{len(genes)=}\")\n",
    "\n",
    "df = pd.DataFrame({'gene_name' : genes})\n",
    "df['gene_type'] = 'transcription_factor'\n",
    "\n",
    "output_path = \"../config/gene_annotations/scenic_transcription_factors.csv\"\n",
    "df.to_csv(output_path, index=False)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07cf149b-fd94-4aa8-8cc2-ec315a55ef46",
   "metadata": {},
   "source": [
    "# Gene Ontology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aec14950-2421-436b-aacf-867150018fd4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hematopoietic stem cell proliferation      32\n",
      "hematopoietic stem cell homeostasis        28\n",
      "hematopoietic stem cell differentiation    25\n",
      "Name: GO NAME, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>go_biological_process</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LMBR1L</td>\n",
       "      <td>hematopoietic stem cell differentiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHD2</td>\n",
       "      <td>hematopoietic stem cell differentiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UFL1</td>\n",
       "      <td>hematopoietic stem cell differentiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UFL1</td>\n",
       "      <td>hematopoietic stem cell differentiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TP53</td>\n",
       "      <td>hematopoietic stem cell differentiation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name                    go_biological_process\n",
       "0    LMBR1L  hematopoietic stem cell differentiation\n",
       "1      CHD2  hematopoietic stem cell differentiation\n",
       "2      UFL1  hematopoietic stem cell differentiation\n",
       "3      UFL1  hematopoietic stem cell differentiation\n",
       "4      TP53  hematopoietic stem cell differentiation"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dpath = \"/nfs/turbo/umms-indikar/shared/projects/HSC/data/GO_annotations/\"\n",
    "\n",
    "hdf = []\n",
    "\n",
    "for f in os.listdir(dpath):\n",
    "    if 'neuronal' in f:\n",
    "        continue\n",
    "        \n",
    "    fpath = f\"{dpath}{f}\"\n",
    "    tmp = pd.read_csv(fpath, sep='\\t')\n",
    "    hdf.append(tmp)\n",
    "\n",
    "hdf = pd.concat(hdf)\n",
    "print(hdf['GO NAME'].value_counts())\n",
    "\n",
    "columns = [\n",
    "    'SYMBOL',\n",
    "    'GO NAME',\n",
    "]\n",
    "\n",
    "hdf = hdf[columns]\n",
    "hdf.columns = [\n",
    "    'gene_name',\n",
    "    'go_biological_process',\n",
    "    \n",
    "]\n",
    "\n",
    "output_path = \"../config/gene_annotations/go_annotations.csv\"\n",
    "hdf.to_csv(output_path, index=False)\n",
    "hdf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0533c46a-0697-40ff-8d5d-f6a9ccfbedbc",
   "metadata": {},
   "source": [
    "# Tabula Sapeins differentially expressed genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "35b83349-8fa5-46d5-81ad-d3f62e23e17a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 27346 × 58870\n",
       "    obs: 'organ_tissue', 'method', 'donor', 'anatomical_information', 'n_counts_UMIs', 'n_genes', 'cell_ontology_class', 'free_annotation', 'manually_annotated', 'compartment', 'gender', 'celltype', 'record_id', 'cell_id'\n",
       "    var: 'gene_symbol', 'feature_type', 'ensemblid', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: '_scvi', '_training_mode', 'dendrogram_cell_type_tissue', 'dendrogram_computational_compartment_assignment', 'dendrogram_consensus_prediction', 'dendrogram_tissue_cell_type', 'donor_colors', 'donor_method_colors', 'hvg', 'log1p', 'method_colors', 'neighbors', 'organ_tissue_colors', 'sex_colors', 'tissue_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_scvi', 'X_scvi_umap', 'X_umap'\n",
       "    layers: 'decontXcounts', 'raw_counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fpath = \"/nfs/turbo/umms-indikar/shared/projects/adaptive_sampling/data/tabula_sapiens_filtered.h5ad\"\n",
    "adata = sc.read_h5ad(fpath)\n",
    "\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fcc48a87-2553-4881-ac1f-65b01a6b3a24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "fibroblast                 26586\n",
       "hematopoietic stem cell      760\n",
       "Name: cell_ontology_class, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['cell_ontology_class'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2058395b-27e8-4270-9ab5-eafebe904fb6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/cstansbu/.local/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:419: RuntimeWarning: overflow encountered in expm1\n",
      "  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (\n",
      "/home/cstansbu/.local/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:420: RuntimeWarning: overflow encountered in expm1\n",
      "  self.expm1_func(mean_rest) + 1e-9\n",
      "/home/cstansbu/.local/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:419: RuntimeWarning: invalid value encountered in divide\n",
      "  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (\n",
      "/home/cstansbu/.local/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:422: RuntimeWarning: divide by zero encountered in log2\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/home/cstansbu/.local/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:419: RuntimeWarning: overflow encountered in expm1\n",
      "  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (\n",
      "/home/cstansbu/.local/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:420: RuntimeWarning: overflow encountered in expm1\n",
      "  self.expm1_func(mean_rest) + 1e-9\n",
      "/home/cstansbu/.local/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:419: RuntimeWarning: invalid value encountered in divide\n",
      "  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>names</th>\n",
       "      <th>scores</th>\n",
       "      <th>logfoldchanges</th>\n",
       "      <th>pvals</th>\n",
       "      <th>pvals_adj</th>\n",
       "      <th>pct_nz_group</th>\n",
       "      <th>pct_nz_reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>fibroblast</td>\n",
       "      <td>DCN</td>\n",
       "      <td>46.815006</td>\n",
       "      <td>874.562866</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.994396</td>\n",
       "      <td>0.003947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fibroblast</td>\n",
       "      <td>MGP</td>\n",
       "      <td>46.606716</td>\n",
       "      <td>445.559448</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.990032</td>\n",
       "      <td>0.002632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>fibroblast</td>\n",
       "      <td>C1S</td>\n",
       "      <td>46.314835</td>\n",
       "      <td>182.270279</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.985406</td>\n",
       "      <td>0.001316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>fibroblast</td>\n",
       "      <td>COL6A2</td>\n",
       "      <td>46.147636</td>\n",
       "      <td>50.370811</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.982322</td>\n",
       "      <td>0.046053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fibroblast</td>\n",
       "      <td>C1R</td>\n",
       "      <td>46.060680</td>\n",
       "      <td>153.544693</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.979576</td>\n",
       "      <td>0.022368</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        group   names     scores  logfoldchanges  pvals  pvals_adj  \\\n",
       "0  fibroblast     DCN  46.815006      874.562866    0.0        0.0   \n",
       "1  fibroblast     MGP  46.606716      445.559448    0.0        0.0   \n",
       "2  fibroblast     C1S  46.314835      182.270279    0.0        0.0   \n",
       "3  fibroblast  COL6A2  46.147636       50.370811    0.0        0.0   \n",
       "4  fibroblast     C1R  46.060680      153.544693    0.0        0.0   \n",
       "\n",
       "   pct_nz_group  pct_nz_reference  \n",
       "0      0.994396          0.003947  \n",
       "1      0.990032          0.002632  \n",
       "2      0.985406          0.001316  \n",
       "3      0.982322          0.046053  \n",
       "4      0.979576          0.022368  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "key = 'cell_ontology_class'\n",
    "\n",
    "sc.tl.rank_genes_groups(adata, \n",
    "                        groupby=key,\n",
    "                        method='wilcoxon',\n",
    "                        key_added=key,\n",
    "                        pts=True,\n",
    "                        corr_method='benjamini-hochberg')\n",
    "\n",
    "deg = sc.get.rank_genes_groups_df(adata, \n",
    "                                  group=None,\n",
    "                                  key=key)\n",
    "\n",
    "deg.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e0483187-c422-4aae-bed4-d9f770a6674d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sig.shape=(3215, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>logfoldchanges</th>\n",
       "      <th>pvals_adj</th>\n",
       "      <th>pct_nz_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>fibroblast</td>\n",
       "      <td>DCN</td>\n",
       "      <td>874.562866</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.994396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fibroblast</td>\n",
       "      <td>MGP</td>\n",
       "      <td>445.559448</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.990032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>fibroblast</td>\n",
       "      <td>C1S</td>\n",
       "      <td>182.270279</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.985406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>fibroblast</td>\n",
       "      <td>COL6A2</td>\n",
       "      <td>50.370811</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.982322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fibroblast</td>\n",
       "      <td>C1R</td>\n",
       "      <td>153.544693</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.979576</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cell_type gene_name  logfoldchanges  pvals_adj  pct_nz_group\n",
       "0  fibroblast       DCN      874.562866        0.0      0.994396\n",
       "1  fibroblast       MGP      445.559448        0.0      0.990032\n",
       "2  fibroblast       C1S      182.270279        0.0      0.985406\n",
       "3  fibroblast    COL6A2       50.370811        0.0      0.982322\n",
       "4  fibroblast       C1R      153.544693        0.0      0.979576"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def filter_dataframe(df, lfc_thresh=2, pval_adj_thresh=0.05, pct_nz_thresh=0.5):\n",
    "  \"\"\"\n",
    "  Filters a DataFrame based on logfoldchanges, pvals_adj, and pct_nz_group.\n",
    "\n",
    "  Args:\n",
    "      df (pandas.DataFrame): The DataFrame to filter.\n",
    "      lfc_thresh (float, optional): Minimum absolute value of logfoldchange. Defaults to 2.\n",
    "      pval_adj_thresh (float, optional): Maximum adjusted p-value. Defaults to 0.05.\n",
    "      pct_nz_thresh (float, optional): Minimum proportion of non-zero values in a group. Defaults to 0.5.\n",
    "\n",
    "  Returns:\n",
    "      pandas.DataFrame: The filtered DataFrame.\n",
    "  \"\"\"\n",
    "\n",
    "  return df[\n",
    "      (df['logfoldchanges'].abs() >= lfc_thresh) &\n",
    "      (df['pvals_adj'] <= pval_adj_thresh) &\n",
    "      (df['pct_nz_group'] >= pct_nz_thresh)\n",
    "  ]\n",
    "\n",
    "sig = filter_dataframe(deg, lfc_thresh=2, pval_adj_thresh=0.05, pct_nz_thresh=0.5)\n",
    "print(f\"{sig.shape=}\")\n",
    "\n",
    "columns = [\n",
    "    'group',\n",
    "    'names', \n",
    "    'logfoldchanges',\n",
    "    'pvals_adj',\n",
    "    'pct_nz_group',\n",
    "]\n",
    "\n",
    "sig = sig[columns]\n",
    "\n",
    "sig.columns = [\n",
    "    'cell_type',\n",
    "    'gene_name',\n",
    "    'logfoldchanges',\n",
    "    'pvals_adj',\n",
    "    'pct_nz_group',\n",
    "]\n",
    "\n",
    "output_path = \"../config/gene_annotations/tabula_sapiens_deg.csv\"\n",
    "sig.to_csv(output_path, index=False)\n",
    "\n",
    "sig.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91e47520-6e81-4259-b39f-8f7bf5ee948e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
