{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7bd54be0-430e-4cd0-99eb-7fe523856e4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "import gget\n",
    "import glob\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.axes_divider import make_axes_locatable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d7f22b2c-b685-4fee-9807-f62ae1bfcec5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gdf.shape=(251121, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENST00000673477</td>\n",
       "      <td>ATAD3B-206</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENST00000472194</td>\n",
       "      <td>ATAD3B-203</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENST00000378736</td>\n",
       "      <td>ATAD3B-202</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENST00000485748</td>\n",
       "      <td>ATAD3B-205</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENST00000474481</td>\n",
       "      <td>ATAD3B-204</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_id gene_name    gene_biotype    transcript_id  \\\n",
       "1   ENSG00000160072    ATAD3B  protein_coding  ENST00000673477   \n",
       "38  ENSG00000160072    ATAD3B  protein_coding  ENST00000472194   \n",
       "53  ENSG00000160072    ATAD3B  protein_coding  ENST00000378736   \n",
       "58  ENSG00000160072    ATAD3B  protein_coding  ENST00000485748   \n",
       "69  ENSG00000160072    ATAD3B  protein_coding  ENST00000474481   \n",
       "\n",
       "   transcript_name    transcript_biotype  \n",
       "1       ATAD3B-206        protein_coding  \n",
       "38      ATAD3B-203       retained_intron  \n",
       "53      ATAD3B-202  processed_transcript  \n",
       "58      ATAD3B-205       retained_intron  \n",
       "69      ATAD3B-204       retained_intron  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def load_gtf(fpath: str = \"/scratch/indikar_root/indikar1/cstansbu/HSC/references/geneTable.csv\") -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Loads a gene table CSV file from the specified path, \n",
    "    cleans the data, and returns a Pandas DataFrame.\n",
    "\n",
    "    Args:\n",
    "        fpath: The path to the CSV file.\n",
    "\n",
    "    Returns:\n",
    "        A cleaned pandas DataFrame containing the gene data.\n",
    "    \"\"\"\n",
    "\n",
    "    columns = [\n",
    "        'transcript_id',\n",
    "        'transcript_name',\n",
    "        'transcript_biotype',\n",
    "        'gene_id',\n",
    "        'gene_name',\n",
    "        'gene_biotype',\n",
    "    ]\n",
    "\n",
    "    # Read only the necessary columns\n",
    "    gdf = pd.read_csv(fpath, low_memory=False, usecols=columns)\n",
    "\n",
    "    # Drop rows with NaN values in key columns and remove duplicates\n",
    "    gdf = gdf.dropna(subset=[\"transcript_id\", \"gene_id\"]).drop_duplicates()\n",
    "\n",
    "    return gdf\n",
    "\n",
    "\n",
    "fpath = \"/scratch/indikar_root/indikar1/cstansbu/HSC/references/geneTable.csv\"\n",
    "gdf = load_gtf(fpath)\n",
    "print(f\"{gdf.shape=}\")\n",
    "gdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "92bcbaa5-5864-4e6e-be00-123f4eb76449",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <td>90.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001460</th>\n",
       "      <td>47.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene_count\n",
       "gene_id                    \n",
       "ENSG00000000457        25.0\n",
       "ENSG00000000460         4.0\n",
       "ENSG00000000938         1.0\n",
       "ENSG00000000971        90.0\n",
       "ENSG00000001460        47.0"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fpath = \"/scratch/indikar_root/indikar1/cstansbu/HSC/isoquant/Gridr3/Gridr3.gene_counts.tsv\"\n",
    "\n",
    "df = pd.read_csv(fpath, \n",
    "                 sep=\"\\t\", \n",
    "                 names=['gene_id', 'gene_count'], \n",
    "                 comment=\"#\")\n",
    "\n",
    "df = df.set_index('gene_id')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35eef162-5896-45b2-8550-f219ba7bcaf5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a70c539-8be3-41e1-bb92-841e36eb489f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b4cb3967-cf0b-4f23-9608-8e9f9d90b383",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "'break' outside loop (668683560.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[5], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    break\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m 'break' outside loop\n"
     ]
    }
   ],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8858c39-ea9a-44e9-9df2-0651edad8808",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2191b83f-7274-491f-844b-3fe8b82e4484",
   "metadata": {},
   "outputs": [],
   "source": [
    "trx = gdf[['transcript_id', 'transcript_name']]\n",
    "trx = trx.drop_duplicates()\n",
    "trx = trx[trx['transcript_id'].notna()]\n",
    "\n",
    "trans_map = dict(zip(trx['transcript_id'].values, trx['transcript_name'].values))\n",
    "\n",
    "trx.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d71f3bb4-3b84-45f0-891b-8a3ffd75e1d8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "132e3f5a-fd8b-4ee5-b508-4886d2f60172",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "227edf38-0828-4d03-a17a-3d3cb4c2d769",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57d6edd6-0d61-460e-97d3-02c437062a52",
   "metadata": {},
   "outputs": [],
   "source": [
    "fpath = \"/scratch/indikar_root/indikar1/cstansbu/HSC/isoquant/Gridr3/Gridr3.transcript_tpm.tsv\"\n",
    "df = pd.read_csv(fpath, sep='\\t')\n",
    "df['transcript_name'] = df['#feature_id'].map(trans_map)\n",
    "df['gene_name'] = df['transcript_name'].apply(lambda x: \"-\".join(str(x).split(\"-\")[:-1]))\n",
    "\n",
    "# drop zeros\n",
    "df = df[df['TPM'] > 0]\n",
    "df['gene_var'] = df.groupby('gene_name')['TPM'].transform('std')\n",
    "df['tx_count'] = df.groupby('gene_name')['TPM'].transform('count')\n",
    "\n",
    "# drop single isoform genes\n",
    "df = df[df['tx_count'] > 1]\n",
    "\n",
    "# drop mitchondrial amnd ribosomal genes\n",
    "df = df[~df['gene_name'].str.startswith(\"MT\")]\n",
    "df = df[~df['gene_name'].str.startswith(\"RP\")]\n",
    "\n",
    "df = df.sort_values(by='gene_var', ascending=False)\n",
    "\n",
    "df.head(20)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7794089a-1759-48ca-87d1-99b78beec31b",
   "metadata": {},
   "source": [
    "# HSC Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bca462c-eed9-4ad4-ad1b-668cfd938e2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = pd.read_csv(\"~/git_repositories/ONT-single-cell/config/gene_annotations/panglaodb.csv\")\n",
    "genes = genes[genes['cell_type'] == 'Hematopoietic stem cells']\n",
    "genes = genes['gene_name'].values\n",
    "print(len(genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9b17764-c875-4443-8336-6203e856494e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf = df.copy()\n",
    "pdf = pdf[pdf['gene_name'].isin(genes)]\n",
    "pdf['gene_TPM'] = pdf.groupby('gene_name')['TPM'].transform('sum')\n",
    "pdf['gene_TPM_log'] = np.log1p(pdf['gene_TPM'])\n",
    "pdf['percent'] = (pdf['TPM'] / pdf['gene_TPM']) * 100\n",
    "print(f\"{pdf.shape=}\")\n",
    "\n",
    "pdf['isoform_name'] = pdf['transcript_name'].apply(lambda x: x.split(\"-\")[1])\n",
    "\n",
    "table = pd.pivot_table(pdf, \n",
    "                     index='gene_name', \n",
    "                     columns='isoform_name',\n",
    "                     values='percent',\n",
    "                     fill_value=0\n",
    "                    )\n",
    "\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 6, 5.25\n",
    "\n",
    "table.plot(kind='bar', \n",
    "           stacked=True,\n",
    "           width=0.75, \n",
    "           cmap='tab20b',\n",
    "           ec='k')\n",
    "\n",
    "\n",
    "sns.move_legend(plt.gca(), \n",
    "                loc='upper right',\n",
    "                title='Isoform',\n",
    "                bbox_to_anchor=(1.2, 1.01))\n",
    "\n",
    "plt.ylabel(\"Percent of Expression\")\n",
    "plt.xlabel(\"\")\n",
    "\n",
    "\n",
    "ax_divider = make_axes_locatable(plt.gca())\n",
    "ax2 = ax_divider.append_axes(\"top\", \n",
    "                             size=\"25%\", \n",
    "                             pad=f\"5%\",\n",
    "                             sharex=plt.gca(),\n",
    "                            )\n",
    "                             \n",
    "\n",
    "# get total expression\n",
    "pdf = pdf[['gene_name', 'gene_TPM_log']].drop_duplicates()\n",
    "pdf = pdf.sort_values(by='gene_name')\n",
    "\n",
    "sns.barplot(data=pdf,\n",
    "            x='gene_name', \n",
    "            y='gene_TPM_log',\n",
    "            ec='k',\n",
    "            width=0.75,\n",
    "            color='lightgrey',\n",
    "            ax=ax2)\n",
    "\n",
    "ax2.set_xticklabels([\"\"]*len(pdf))\n",
    "plt.xlabel(\"\")\n",
    "plt.ylabel('TPM (log)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da6763ea-b3eb-4608-9b8d-cbf8b07da4f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf = pdf[['gene_name', 'gene_TPM']].drop_duplicates()\n",
    "pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44fb80e8-3606-47e4-a4e1-36ff5c59aa80",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinf",
   "language": "python",
   "name": "bioinf"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
