{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e396b914-ab5b-48a5-8614-d4956e20aeb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "\n",
    "import anndata as ad\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a98be0de-29c6-4fb3-80ac-8c974c1f933e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gf.shape=(3371244, 26)\n",
      "gf.shape=(20023, 26)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>1</td>\n",
       "      <td>1471764</td>\n",
       "      <td>1497848</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000142611</td>\n",
       "      <td>PRDM16</td>\n",
       "      <td>1</td>\n",
       "      <td>3069167</td>\n",
       "      <td>3438621</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000157933</td>\n",
       "      <td>SKI</td>\n",
       "      <td>1</td>\n",
       "      <td>2228318</td>\n",
       "      <td>2310213</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000142655</td>\n",
       "      <td>PEX14</td>\n",
       "      <td>1</td>\n",
       "      <td>10472287</td>\n",
       "      <td>10630758</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000149527</td>\n",
       "      <td>PLCH2</td>\n",
       "      <td>1</td>\n",
       "      <td>2425979</td>\n",
       "      <td>2505532</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id gene_name Chromosome     Start       End Strand\n",
       "0  ENSG00000160072    ATAD3B          1   1471764   1497848      +\n",
       "1  ENSG00000142611    PRDM16          1   3069167   3438621      +\n",
       "2  ENSG00000157933       SKI          1   2228318   2310213      +\n",
       "3  ENSG00000142655     PEX14          1  10472287  10630758      +\n",
       "4  ENSG00000149527     PLCH2          1   2425979   2505532      +"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fpath = '/scratch/indikar_root/indikar1/cstansbu/sc-iHSC/references/geneTable.csv'\n",
    "gf = pd.read_csv(fpath, low_memory=False)\n",
    "print(f\"{gf.shape=}\")\n",
    "\n",
    "# subset just the major genes\n",
    "gf = gf[gf['Feature'] == 'gene']\n",
    "gf = gf[gf['gene_biotype'] == 'protein_coding'].reset_index(drop=True)\n",
    "print(f\"{gf.shape=}\")\n",
    "\n",
    "keep_cols = [\n",
    "    'gene_id', \n",
    "    'gene_name',\n",
    "    'Chromosome', \n",
    "    'Start',\n",
    "    'End',\n",
    "    'Strand',\n",
    "]\n",
    "\n",
    "gf = gf[keep_cols]\n",
    "gf = gf[gf['gene_name'].notna()]\n",
    "gf = gf.drop_duplicates(subset='gene_name')\n",
    "gf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bd53b52c-1000-47d8-91c0-b58e767aec82",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df.shape=(61865, 8526)\n",
      "df.shape=(19393, 8524) gf.shape=(19393, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_id</th>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <th>ENSG00000001084</th>\n",
       "      <th>ENSG00000001167</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000288825</th>\n",
       "      <th>ENSG00000288859</th>\n",
       "      <th>ENSG00000288864</th>\n",
       "      <th>ENSG00000288867</th>\n",
       "      <th>ENSG00000288920</th>\n",
       "      <th>ENSG00000289549</th>\n",
       "      <th>ENSG00000289604</th>\n",
       "      <th>ENSG00000289716</th>\n",
       "      <th>ENSG00000289721</th>\n",
       "      <th>ENSG00000289746</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGGTTACCT</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGTTGAAGT</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGTTGTCGT</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAGAAGCGT</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAGGAGGTT</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 19393 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_id           ENSG00000000003  ENSG00000000005  ENSG00000000419  \\\n",
       "AAACCCAAGGTTACCT                0                0                0   \n",
       "AAACCCAAGTTGAAGT                0                0                0   \n",
       "AAACCCAAGTTGTCGT                0                0                0   \n",
       "AAACCCACAGAAGCGT                0                0                0   \n",
       "AAACCCACAGGAGGTT                0                0                0   \n",
       "\n",
       "gene_id           ENSG00000000457  ENSG00000000460  ENSG00000000938  \\\n",
       "AAACCCAAGGTTACCT                0                0                0   \n",
       "AAACCCAAGTTGAAGT                0                0                0   \n",
       "AAACCCAAGTTGTCGT                0                0                0   \n",
       "AAACCCACAGAAGCGT                0                0                0   \n",
       "AAACCCACAGGAGGTT                0                0                0   \n",
       "\n",
       "gene_id           ENSG00000000971  ENSG00000001036  ENSG00000001084  \\\n",
       "AAACCCAAGGTTACCT                0                0                0   \n",
       "AAACCCAAGTTGAAGT                0                0                0   \n",
       "AAACCCAAGTTGTCGT                0                0                0   \n",
       "AAACCCACAGAAGCGT                0                0                0   \n",
       "AAACCCACAGGAGGTT                0                0                0   \n",
       "\n",
       "gene_id           ENSG00000001167  ...  ENSG00000288825  ENSG00000288859  \\\n",
       "AAACCCAAGGTTACCT                0  ...                0                0   \n",
       "AAACCCAAGTTGAAGT                0  ...                0                0   \n",
       "AAACCCAAGTTGTCGT                0  ...                0                0   \n",
       "AAACCCACAGAAGCGT                0  ...                0                0   \n",
       "AAACCCACAGGAGGTT                0  ...                0                0   \n",
       "\n",
       "gene_id           ENSG00000288864  ENSG00000288867  ENSG00000288920  \\\n",
       "AAACCCAAGGTTACCT                0                0                0   \n",
       "AAACCCAAGTTGAAGT                0                0                0   \n",
       "AAACCCAAGTTGTCGT                0                0                0   \n",
       "AAACCCACAGAAGCGT                0                0                0   \n",
       "AAACCCACAGGAGGTT                0                0                0   \n",
       "\n",
       "gene_id           ENSG00000289549  ENSG00000289604  ENSG00000289716  \\\n",
       "AAACCCAAGGTTACCT                0                0                0   \n",
       "AAACCCAAGTTGAAGT                0                0                0   \n",
       "AAACCCAAGTTGTCGT                0                0                0   \n",
       "AAACCCACAGAAGCGT                0                0                0   \n",
       "AAACCCACAGGAGGTT                0                0                0   \n",
       "\n",
       "gene_id           ENSG00000289721  ENSG00000289746  \n",
       "AAACCCAAGGTTACCT                0                0  \n",
       "AAACCCAAGTTGAAGT                0                0  \n",
       "AAACCCAAGTTGTCGT                0                0  \n",
       "AAACCCACAGAAGCGT                0                0  \n",
       "AAACCCACAGGAGGTT                0                0  \n",
       "\n",
       "[5 rows x 19393 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fpath = '/scratch/indikar_root/indikar1/cstansbu/sc-iHSC/counts/counts.txt'\n",
    "\n",
    "df = pd.read_csv(fpath, sep='\\t', low_memory=False)\n",
    "print(f\"{df.shape=}\")\n",
    "\n",
    "df = df.rename(columns={'Unnamed: 0' : 'gene_id', 'Unnamed: 1' : 'chromosome'})\n",
    "df = df.drop(columns='chromosome')\n",
    "\n",
    "# filter out non-PT genes\n",
    "df = df[df['gene_id'].isin(gf['gene_id'])]\n",
    "df = df.set_index('gene_id')\n",
    "\n",
    "# make sure that the GTF has the same set\n",
    "gf = gf[gf['gene_id'].isin(df.index)]\n",
    "\n",
    "gf = gf.set_index('gene_id')\n",
    "\n",
    "print(f\"{df.shape=} {gf.shape=}\")\n",
    "\n",
    "##TRANSPOSE \n",
    "df = df.T\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f0e1cf77-1525-41c9-9666-6d44c6faacc0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 8524 × 19393\n",
       "    obs: 'n_genes'\n",
       "    var: 'gene_name', 'Chromosome', 'Start', 'End', 'Strand'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create the anndata object\n",
    "\n",
    "adata = ad.AnnData(df.to_numpy(), dtype='int64')\n",
    "adata.obs_names = df.index\n",
    "adata.var_names = df.columns\n",
    "adata.var = gf\n",
    "\n",
    "obs = df.sum(axis=1).reset_index(drop=False)\n",
    "obs.columns = ['cell_id', 'n_genes']\n",
    "obs = obs.set_index('cell_id')\n",
    "\n",
    "adata.obs = obs\n",
    "\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e9b38e3e-5c73-4e98-b1b2-a8cca9b98be4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGGTTACCT</th>\n",
       "      <td>352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGTTGAAGT</th>\n",
       "      <td>276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGTTGTCGT</th>\n",
       "      <td>225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAGAAGCGT</th>\n",
       "      <td>194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAGGAGGTT</th>\n",
       "      <td>787</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  n_genes\n",
       "cell_id                  \n",
       "AAACCCAAGGTTACCT      352\n",
       "AAACCCAAGTTGAAGT      276\n",
       "AAACCCAAGTTGTCGT      225\n",
       "AAACCCACAGAAGCGT      194\n",
       "AAACCCACAGGAGGTT      787"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa097b48-2a36-41e2-ad13-241aac34444b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
