{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "071b7b26-5cdd-4b44-8ce9-0d81bdc4e5ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import scipy.sparse as sp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2a98a801-e3d2-4fe2-92b1-3b0751f60e6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 34004 × 58870\n",
       "    obs: 'organ_tissue', 'method', 'donor', 'anatomical_information', 'n_counts_UMIs', 'n_genes', 'cell_ontology_class', 'free_annotation', 'manually_annotated', 'compartment', 'gender'\n",
       "    var: 'gene_symbol', 'feature_type', 'ensemblid', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: '_scvi', '_training_mode', 'cell_ontology_class_colors', 'dendrogram_cell_type_tissue', 'dendrogram_computational_compartment_assignment', 'dendrogram_consensus_prediction', 'dendrogram_tissue_cell_type', 'donor_colors', 'donor_method_colors', 'hvg', 'method_colors', 'neighbors', 'organ_tissue_colors', 'sex_colors', 'tissue_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_scvi', 'X_scvi_umap', 'X_umap'\n",
       "    layers: 'decontXcounts', 'raw_counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fpath = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Spleen.h5ad\"\n",
    "\n",
    "adata = sc.read_h5ad(fpath)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "93d86f29-6ff0-44f2-baa0-957fef2841a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len(cell_mask)=34004\n",
      "len(gene_mask)=49340\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 596 × 49340\n",
       "    obs: 'organ_tissue', 'method', 'donor', 'anatomical_information', 'n_counts_UMIs', 'n_genes', 'cell_ontology_class', 'free_annotation', 'manually_annotated', 'compartment', 'gender'\n",
       "    var: 'gene_symbol', 'feature_type', 'ensemblid', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: '_scvi', '_training_mode', 'cell_ontology_class_colors', 'dendrogram_cell_type_tissue', 'dendrogram_computational_compartment_assignment', 'dendrogram_consensus_prediction', 'dendrogram_tissue_cell_type', 'donor_colors', 'donor_method_colors', 'hvg', 'method_colors', 'neighbors', 'organ_tissue_colors', 'sex_colors', 'tissue_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_scvi', 'X_scvi_umap', 'X_umap'\n",
       "    layers: 'decontXcounts', 'raw_counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get all cells of given type\n",
    "cell_type = 'endothelial cell'\n",
    "cell_mask = (adata.obs['free_annotation'] == cell_type)\n",
    "\n",
    "# get non-zero genes\n",
    "nonzero_indices = np.argwhere(np.ravel(adata.layers['raw_counts'].sum(axis=0)) > 0)\n",
    "nonzero_indices = np.ravel(nonzero_indices)\n",
    "genes_to_keep = adata.var.iloc[nonzero_indices]\n",
    "\n",
    "# drop mitochondrial and ribosomal genes\n",
    "genes_to_keep = genes_to_keep[~genes_to_keep['gene_symbol'].str.startswith('MT')]\n",
    "genes_to_keep = genes_to_keep[~genes_to_keep['gene_symbol'].str.startswith('RP')]\n",
    "\n",
    "gene_mask = genes_to_keep.index\n",
    "\n",
    "print(f\"{len(cell_mask)=}\")\n",
    "print(f\"{len(gene_mask)=}\")\n",
    "\n",
    "# extract the data \n",
    "pdf = adata[cell_mask, gene_mask]\n",
    "pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "574d5d88-0ec2-41ea-bb58-3f028ca8a36f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_id</th>\n",
       "      <th>DDX11L1</th>\n",
       "      <th>WASH7P</th>\n",
       "      <th>MIR6859-1</th>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <th>OR4F5</th>\n",
       "      <th>AL627309.1</th>\n",
       "      <th>AL627309.3</th>\n",
       "      <th>CICP27</th>\n",
       "      <th>AL627309.6</th>\n",
       "      <th>...</th>\n",
       "      <th>AC006386.2</th>\n",
       "      <th>LINC00266-4P</th>\n",
       "      <th>CICP1</th>\n",
       "      <th>AC007562.1</th>\n",
       "      <th>AC024067.1</th>\n",
       "      <th>RF00156-25</th>\n",
       "      <th>ANKRD36P1</th>\n",
       "      <th>PARP4P1</th>\n",
       "      <th>CCNQP2</th>\n",
       "      <th>SPRY3-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCATCCGTGTCT_TSP7_Spleen_NA_10X_1_1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAGGTAAGGTCATAA_TSP7_Spleen_NA_10X_1_1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AACAGGGAGTGCCGAA_TSP7_Spleen_NA_10X_1_1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AACCAACGTAGAGATT_TSP7_Spleen_NA_10X_1_1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AATCACGCAACTCGAT_TSP7_Spleen_NA_10X_1_1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 49341 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   cell_id  DDX11L1  WASH7P  MIR6859-1  \\\n",
       "0  AAACCCATCCGTGTCT_TSP7_Spleen_NA_10X_1_1        0       0          0   \n",
       "1  AAAGGTAAGGTCATAA_TSP7_Spleen_NA_10X_1_1        0       0          0   \n",
       "2  AACAGGGAGTGCCGAA_TSP7_Spleen_NA_10X_1_1        0       0          0   \n",
       "3  AACCAACGTAGAGATT_TSP7_Spleen_NA_10X_1_1        0       0          0   \n",
       "4  AATCACGCAACTCGAT_TSP7_Spleen_NA_10X_1_1        0       0          0   \n",
       "\n",
       "   MIR1302-2HG  OR4F5  AL627309.1  AL627309.3  CICP27  AL627309.6  ...  \\\n",
       "0            0      0           0           0       0           0  ...   \n",
       "1            0      0           0           0       0           0  ...   \n",
       "2            0      0           0           0       0           0  ...   \n",
       "3            0      0           0           0       0           0  ...   \n",
       "4            0      0           0           0       0           0  ...   \n",
       "\n",
       "   AC006386.2  LINC00266-4P  CICP1  AC007562.1  AC024067.1  RF00156-25  \\\n",
       "0           0             0      0           0           0           0   \n",
       "1           0             0      0           0           0           0   \n",
       "2           0             0      0           0           0           0   \n",
       "3           0             0      0           0           0           0   \n",
       "4           0             0      0           0           0           0   \n",
       "\n",
       "   ANKRD36P1  PARP4P1  CCNQP2  SPRY3-1  \n",
       "0          0        0       0        0  \n",
       "1          0        0       0        0  \n",
       "2          0        0       0        0  \n",
       "3          0        0       0        0  \n",
       "4          0        0       0        0  \n",
       "\n",
       "[5 rows x 49341 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pdf.to_df(layer='raw_counts')\n",
    "outpath = \"/home/cstansbu/temp/tabula_sapeins_endothelial_cell.csv\" \n",
    "df = df.astype(int)\n",
    "df = df.reset_index()\n",
    "df.to_csv(outpath, index=False)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6937c9ef-f2b4-4f5a-8fca-7981b157a16c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "80518"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.set_index('cell_id').to_numpy().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "584abcca-c1db-48d3-9f32-f11fff56aa99",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97f44155-1fa7-499d-a707-d8555b812a46",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec74e858-7efc-42a7-94f4-887b5a88305f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d33e819e-8ab0-46e2-8856-310cc29f5611",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nonzero_median(df):\n",
    "    \"\"\"\n",
    "    Calculates the median of nonzero values for each column in a Pandas DataFrame.\n",
    "\n",
    "    Args:\n",
    "        df (pandas.DataFrame): The DataFrame to compute medians for.\n",
    "\n",
    "    Returns:\n",
    "        pandas.Series: A Series containing the median of nonzero values for each column.\n",
    "    \"\"\"\n",
    "    # Replace zeros with NaN\n",
    "    df_nonzero = df.replace(0, np.nan) \n",
    "\n",
    "    # Calculate medians, skipping NaNs (which were the zeros)\n",
    "    medians = df_nonzero.median(skipna=True)  \n",
    "    return medians\n",
    "\n",
    "\n",
    "def normalize_counts(X: sp.csr_matrix, \n",
    "                     norm_factor_vector: np.ndarray, \n",
    "                     n_counts: np.ndarray, \n",
    "                     target_sum: int = 10000) -> sp.csr_matrix:\n",
    "    \"\"\"\n",
    "    Normalizes and scales count data in a sparse matrix.\n",
    "\n",
    "    This function is commonly used for library size normalization and scaling \n",
    "    in single-cell RNA sequencing analysis. It adjusts raw counts to account \n",
    "    for differences in sequencing depth and scales them to a target sum \n",
    "    (e.g., CPM or TPM normalization).\n",
    "\n",
    "    Args:\n",
    "        X: A sparse matrix (csr_matrix) where rows are features (e.g., genes) \n",
    "           and columns are samples (e.g., cells). Values represent raw counts.\n",
    "        norm_factor_vector: A 1D array containing the normalization factor for \n",
    "                             each sample (e.g., library size, sequencing depth).\n",
    "        n_counts: A 1D array containing the total count per sample.\n",
    "        target_sum: The desired sum of normalized counts per sample. \n",
    "                     Defaults to 10,000.\n",
    "\n",
    "    Returns:\n",
    "        A sparse matrix (csr_matrix) with normalized and scaled counts.\n",
    "\n",
    "    Raises:\n",
    "        ValueError: If dimensions of `X`, `norm_factor_vector`, and `n_counts`\n",
    "                    are incompatible.\n",
    "    \"\"\"\n",
    "    X_norm = (X / n_counts) * target_sum / norm_factor_vector\n",
    "    return sp.csr_matrix(X_norm)\n",
    "\n",
    "\n",
    "n_cells = 1000\n",
    "X = df.sample(n_cells)\n",
    "X = X.set_index('cell_id')\n",
    "\n",
    "# row sums \n",
    "n_counts = X.sum(axis=1)\n",
    "\n",
    "# nonzero column medians - needs work\n",
    "norm_factor_vector = nonzero_median(X)\n",
    "\n",
    "X_norm = normalize_counts(X, norm_factor_vector, n_counts)\n",
    "X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "552ff12d-263a-4cf6-b219-fd457c269cd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0199b27-ba32-4d69-b4b0-65b220e800c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "A = \"cooper\"\n",
    "?A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea704585-24a8-4440-8ca8-1dea72a68ea3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9e39ab7-d64c-4d82-9bf4-fbc5cd5e6409",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e625d479-1313-454b-bc83-db5b240196ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "?pd.DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11ab7fa4-6c63-4a76-9c5f-adb0da941659",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "scanpy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
