{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3d8f9f1e-2f69-4432-bc42-97800e10a5e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import subprocess\n",
    "import shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pysam\n",
    "import json\n",
    "import glob\n",
    "import re\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "94e09cf5-a785-462f-ac3a-4e179d9d0f49",
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_date_string(date_str):\n",
    "  \"\"\"\n",
    "  This function takes a string in various date formats and returns a consistently formatted YYYY-MM-DD string.\n",
    "\n",
    "  Args:\n",
    "      date_str (str): The input date string.\n",
    "\n",
    "  Returns:\n",
    "      str: The consistently formatted date string (YYYY-MM-DD).\n",
    "      None: If the input string does not contain a valid date.\n",
    "  \"\"\"\n",
    "\n",
    "  # Attempt to match YYYY-MM-DD at the beginning\n",
    "  match = re.match(r'^(\\d{4})-(\\d{2})-(\\d{2})', date_str)\n",
    "  if match:\n",
    "      return '-'.join(match.groups())\n",
    "\n",
    "  # Otherwise, try to extract date components from other parts\n",
    "  for pattern in (r'\\d{8}', r'\\d{6}'):  # Patterns for MMDDYYYY and MMDDYY\n",
    "      match = re.search(pattern, date_str)\n",
    "      if match:\n",
    "          date = match.group(0)\n",
    "          year = date[4:] if len(date) == 8 else '20' + date[4:]\n",
    "          month = date[:2]\n",
    "          day = date[2:4]\n",
    "          return f\"{year}-{month}-{day}\"\n",
    "\n",
    "  # If no valid date found\n",
    "  return None \n",
    "\n",
    "def parse_config(input_list):\n",
    "  \"\"\"\n",
    "  This function parses a list of command-line arguments (strings) into a dictionary.\n",
    "\n",
    "  Args:\n",
    "      input_list (list): A list of strings representing command-line arguments.\n",
    "\n",
    "  Returns:\n",
    "      dict: A dictionary containing the parsed key-value pairs.\n",
    "  \"\"\"\n",
    "  config_dict = {}\n",
    "  for arg in input_list:\n",
    "    if '=' in arg:\n",
    "      key, value = arg.split('=')\n",
    "      # Remove leading/trailing quotes if present around the value\n",
    "      value = value.strip('\"')\n",
    "      # Handle values that may contain spaces (e.g., file paths)\n",
    "      if value.startswith('[') and value.endswith(']'):\n",
    "        try:\n",
    "          # Try parsing list values (e.g., reference_files)\n",
    "          value = eval(value)\n",
    "        except (NameError, SyntaxError):\n",
    "          # If parsing fails, treat it as a string\n",
    "          pass\n",
    "      config_dict[key.strip('--')] = value\n",
    "  return config_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "52eb63ee-4056-4908-9d69-983d01b9ab4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/nfs/turbo/umms-indikar/shared/projects/HSC/data/adaptive_sampling/experiment_data/2024-04-26-iHSC-AS-6th-Run/report_FAX16607_20240426_1111_e8a5d8e8.json\n",
      "/nfs/turbo/umms-indikar/shared/projects/HSC/data/adaptive_sampling/experiment_data/fifth_AS_test_03142024/report_PAS55331_20240314_1638_bfc3b4f2.json\n",
      "/nfs/turbo/umms-indikar/shared/projects/HSC/data/adaptive_sampling/experiment_data/first_AS_test_03122024/report_PAS55331_20240312_1414_1cbba94d.json\n",
      "/nfs/turbo/umms-indikar/shared/projects/HSC/data/adaptive_sampling/experiment_data/sixth_AS_test_03142024/report_PAS55331_20240314_1654_9357e86a.json\n",
      "/nfs/turbo/umms-indikar/shared/projects/HSC/data/adaptive_sampling/experiment_data/2024-04-26-iHSC-AS-7th-Run/report_FAX16607_20240426_1531_0e7ed665.json\n",
      "/nfs/turbo/umms-indikar/shared/projects/HSC/data/adaptive_sampling/experiment_data/second_AS_test_03132024/report_PAS55331_20240313_1306_36a181e6.json\n",
      "/nfs/turbo/umms-indikar/shared/projects/HSC/data/adaptive_sampling/experiment_data/seventh_AS_03152024/report_PAS55331_20240315_1631_f87da033.json\n",
      "/nfs/turbo/umms-indikar/shared/projects/HSC/data/adaptive_sampling/experiment_data/second_iHSC_AS_04162024/report_FAX16607_20240416_1223_a3048d6a.json\n",
      "/nfs/turbo/umms-indikar/shared/projects/HSC/data/adaptive_sampling/experiment_data/2024-04-26-iHSC-AS-8th-Run-fast/report_FAX16607_20240426_1636_942aaa97.json\n",
      "/nfs/turbo/umms-indikar/shared/projects/HSC/data/adaptive_sampling/experiment_data/fourth_AS_test_03142024/report_PAS55331_20240314_1258_65c40cb7.json\n",
      "/nfs/turbo/umms-indikar/shared/projects/HSC/data/adaptive_sampling/experiment_data/third_AS_test_03142024/report_PAS55331_20240314_1244_da7b98e5.json\n",
      "/nfs/turbo/umms-indikar/shared/projects/HSC/data/adaptive_sampling/experiment_data/first_iHSC_AS_04152024/report_FAX16607_20240415_1419_9d766fe6.json\n",
      "/nfs/turbo/umms-indikar/shared/projects/HSC/data/adaptive_sampling/experiment_data/2024-04-22-iHSC-4th-run/report_FAX16607_20240422_1014_2030f100.json\n",
      "/nfs/turbo/umms-indikar/shared/projects/HSC/data/adaptive_sampling/experiment_data/2024-05-03-iHSC-AS-9th-run/report_FAX16607_20240503_1007_f3a56365.json\n",
      "/nfs/turbo/umms-indikar/shared/projects/HSC/data/adaptive_sampling/experiment_data/2024-04-18-iHSC-3rd-run/report_FAX16607_20240418_1022_52a9eff7.json\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_name</th>\n",
       "      <th>date</th>\n",
       "      <th>device</th>\n",
       "      <th>server_path</th>\n",
       "      <th>flow_cell_id</th>\n",
       "      <th>product_code</th>\n",
       "      <th>flowcell_use_count</th>\n",
       "      <th>kit</th>\n",
       "      <th>config_file</th>\n",
       "      <th>fast5</th>\n",
       "      <th>...</th>\n",
       "      <th>guppy_filename</th>\n",
       "      <th>min_qscore</th>\n",
       "      <th>min_read_length</th>\n",
       "      <th>split_files_by_barcode</th>\n",
       "      <th>pod5_reads_per_file</th>\n",
       "      <th>fastq_reads_per_file</th>\n",
       "      <th>raw</th>\n",
       "      <th>bam_only_write_primary_alignment</th>\n",
       "      <th>fast5_reads_per_file</th>\n",
       "      <th>enable</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2024-04-26-iHSC-AS-6th-Run</td>\n",
       "      <td>2024-04-26</td>\n",
       "      <td>GRIDION</td>\n",
       "      <td>2024-04-26-iHSC-AS-6th-Run</td>\n",
       "      <td>FAX16607</td>\n",
       "      <td>FLO-MIN106</td>\n",
       "      <td>5</td>\n",
       "      <td>SQK-PCS111</td>\n",
       "      <td>custom_MIN106_DNA.toml</td>\n",
       "      <td>on</td>\n",
       "      <td>...</td>\n",
       "      <td>dna_r9.4.1_450bps_hac.cfg</td>\n",
       "      <td>9</td>\n",
       "      <td>200</td>\n",
       "      <td>off</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fifth_AS_test_03142024</td>\n",
       "      <td>2024-03-14</td>\n",
       "      <td>P2_SOLO</td>\n",
       "      <td>2024-03-14-Adaptive-Sampling-05</td>\n",
       "      <td>PAS55331</td>\n",
       "      <td>FLO-PRO114M</td>\n",
       "      <td>7</td>\n",
       "      <td>SQK-LSK114</td>\n",
       "      <td>custom_PRO114_DNA.toml</td>\n",
       "      <td>off</td>\n",
       "      <td>...</td>\n",
       "      <td>dna_r10.4.1_e8.2_400bps_5khz_fast_prom.cfg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4000</td>\n",
       "      <td>4000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>first_AS_test_03122024</td>\n",
       "      <td>2024-03-12</td>\n",
       "      <td>P2_SOLO</td>\n",
       "      <td>2024-03-12-iHSc-Adaptive-Sequencing-01</td>\n",
       "      <td>PAS55331</td>\n",
       "      <td>FLO-PRO114</td>\n",
       "      <td>3</td>\n",
       "      <td>SQK-LSK114</td>\n",
       "      <td>custom_PRO114_DNA.toml</td>\n",
       "      <td>off</td>\n",
       "      <td>...</td>\n",
       "      <td>dna_r10.4.1_e8.2_400bps_5khz_fast_prom.cfg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4000</td>\n",
       "      <td>4000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sixth_AS_test_03142024</td>\n",
       "      <td>2024-03-14</td>\n",
       "      <td>P2_SOLO</td>\n",
       "      <td>2024-03-14-06</td>\n",
       "      <td>PAS55331</td>\n",
       "      <td>FLO-PRO114M</td>\n",
       "      <td>8</td>\n",
       "      <td>SQK-LSK114</td>\n",
       "      <td>custom_PRO114_DNA.toml</td>\n",
       "      <td>off</td>\n",
       "      <td>...</td>\n",
       "      <td>dna_r10.4.1_e8.2_400bps_5khz_fast_prom.cfg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4000</td>\n",
       "      <td>4000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2024-04-26-iHSC-AS-7th-Run</td>\n",
       "      <td>2024-04-26</td>\n",
       "      <td>GRIDION</td>\n",
       "      <td>2024-04-26-iHSC-7th-Run</td>\n",
       "      <td>FAX16607</td>\n",
       "      <td>FLO-MIN106</td>\n",
       "      <td>7</td>\n",
       "      <td>SQK-PCS111</td>\n",
       "      <td>custom_MIN106_DNA.toml</td>\n",
       "      <td>on</td>\n",
       "      <td>...</td>\n",
       "      <td>dna_r9.4.1_450bps_hac.cfg</td>\n",
       "      <td>9</td>\n",
       "      <td>200</td>\n",
       "      <td>off</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     run_name        date   device  \\\n",
       "0  2024-04-26-iHSC-AS-6th-Run  2024-04-26  GRIDION   \n",
       "1      fifth_AS_test_03142024  2024-03-14  P2_SOLO   \n",
       "2      first_AS_test_03122024  2024-03-12  P2_SOLO   \n",
       "3      sixth_AS_test_03142024  2024-03-14  P2_SOLO   \n",
       "4  2024-04-26-iHSC-AS-7th-Run  2024-04-26  GRIDION   \n",
       "\n",
       "                              server_path flow_cell_id product_code  \\\n",
       "0              2024-04-26-iHSC-AS-6th-Run     FAX16607   FLO-MIN106   \n",
       "1         2024-03-14-Adaptive-Sampling-05     PAS55331  FLO-PRO114M   \n",
       "2  2024-03-12-iHSc-Adaptive-Sequencing-01     PAS55331   FLO-PRO114   \n",
       "3                           2024-03-14-06     PAS55331  FLO-PRO114M   \n",
       "4                 2024-04-26-iHSC-7th-Run     FAX16607   FLO-MIN106   \n",
       "\n",
       "   flowcell_use_count         kit             config_file fast5  ...  \\\n",
       "0                   5  SQK-PCS111  custom_MIN106_DNA.toml    on  ...   \n",
       "1                   7  SQK-LSK114  custom_PRO114_DNA.toml   off  ...   \n",
       "2                   3  SQK-LSK114  custom_PRO114_DNA.toml   off  ...   \n",
       "3                   8  SQK-LSK114  custom_PRO114_DNA.toml   off  ...   \n",
       "4                   7  SQK-PCS111  custom_MIN106_DNA.toml    on  ...   \n",
       "\n",
       "                               guppy_filename min_qscore min_read_length  \\\n",
       "0                   dna_r9.4.1_450bps_hac.cfg          9             200   \n",
       "1  dna_r10.4.1_e8.2_400bps_5khz_fast_prom.cfg        NaN             200   \n",
       "2  dna_r10.4.1_e8.2_400bps_5khz_fast_prom.cfg        NaN             200   \n",
       "3  dna_r10.4.1_e8.2_400bps_5khz_fast_prom.cfg        NaN             200   \n",
       "4                   dna_r9.4.1_450bps_hac.cfg          9             200   \n",
       "\n",
       "  split_files_by_barcode pod5_reads_per_file fastq_reads_per_file  raw  \\\n",
       "0                    off                 NaN                  NaN  NaN   \n",
       "1                    NaN                4000                 4000  NaN   \n",
       "2                    NaN                4000                 4000  NaN   \n",
       "3                    NaN                4000                 4000  NaN   \n",
       "4                    off                 NaN                  NaN  NaN   \n",
       "\n",
       "  bam_only_write_primary_alignment fast5_reads_per_file enable  \n",
       "0                              NaN                  NaN    NaN  \n",
       "1                              NaN                  NaN    NaN  \n",
       "2                              NaN                  NaN    NaN  \n",
       "3                              NaN                  NaN    NaN  \n",
       "4                              NaN                  NaN    NaN  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dpath = \"/nfs/turbo/umms-indikar/shared/projects/HSC/data/adaptive_sampling/experiment_data/\"\n",
    "\n",
    "res = []\n",
    "\n",
    "for subdir in os.listdir(dpath):\n",
    "    date = format_date_string(subdir)\n",
    "    \n",
    "\n",
    "\n",
    "    search = f\"{dpath}{subdir}/report*.json\"\n",
    "    report_path = glob.glob(search)[0]\n",
    "    print(report_path)\n",
    "\n",
    "    with open(report_path, 'r') as file:\n",
    "        data = json.load(file)\n",
    "\n",
    "        if 'use_count' in data['protocol_run_info']['flow_cell'].keys():\n",
    "            use_count = data['protocol_run_info']['flow_cell']['use_count']\n",
    "        else:\n",
    "            use_count\n",
    "\n",
    "        row = {\n",
    "            'run_name' : subdir,\n",
    "            'date' : date,\n",
    "            'device' : data['protocol_run_info']['device']['device_type'],\n",
    "            'server_path' : data['protocol_run_info']['user_info']['protocol_group_id'],\n",
    "            'flow_cell_id' : data['protocol_run_info']['user_info']['user_specified_flow_cell_id'],\n",
    "            'product_code' : data['protocol_run_info']['user_info']['user_specified_product_code'],\n",
    "            'flowcell_use_count' : use_count,\n",
    "            'kit' : data['protocol_run_info']['meta_info']['tags']['kit']['string_value'],\n",
    "            'config_file' : data['protocol_run_info']['meta_info']['tags']['config path']['string_value'].split(\"/\")[-1],\n",
    "        }\n",
    "\n",
    "        row.update(parse_config(data['protocol_run_info']['args']))\n",
    "        res.append(row)\n",
    "\n",
    "res = pd.DataFrame(res)\n",
    "res.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "af475a39-715b-42f6-8c6b-6918528034ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = res.sort_values(by='run_name')\n",
    "res.to_csv('metadata.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c863a530-66ef-44be-9f70-8acb400040bc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1bd0cf40-798b-4d69-ab7a-09e52f5793e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['--fast5=on',\n",
       " '--pod5=off',\n",
       " '--fastq=on',\n",
       " '--bam=on',\n",
       " '--bam_only_write_primary_alignment=off',\n",
       " '--generate_bulk_file=on',\n",
       " '--active_channel_selection=on',\n",
       " '--base_calling=on',\n",
       " '--read_until',\n",
       " 'reference_files=[\"/data/ADAPTIVE_SAMPLING_iHSC/2000_closest.fasta\"]',\n",
       " 'filter_type=enrich',\n",
       " 'first_channel=1',\n",
       " 'last_channel=512',\n",
       " '--fast5_reads_per_file=4000',\n",
       " '--fast5_data',\n",
       " 'vbz_compress',\n",
       " '--fastq_reads_per_file=4000',\n",
       " '--fastq_data',\n",
       " 'compress',\n",
       " '--bulk_file_content',\n",
       " 'events=[[1,512]]',\n",
       " 'read_table=[[1,512]]',\n",
       " '--mux_scan_period=1.5',\n",
       " '--pore_reserve=on',\n",
       " '--guppy_filename=dna_r9.4.1_450bps_hac.cfg',\n",
       " '--alignment',\n",
       " 'reference_files=[\"/data/ADAPTIVE_SAMPLING_iHSC/2000_closest.fasta\"]',\n",
       " '--min_read_length=200']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t = data['protocol_run_info']['args']\n",
    "t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4c7228a1-98df-497c-8364-f526a9ce24b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'fast5': 'on',\n",
       " 'pod5': 'off',\n",
       " 'fastq': 'on',\n",
       " 'bam': 'on',\n",
       " 'bam_only_write_primary_alignment': 'off',\n",
       " 'generate_bulk_file': 'on',\n",
       " 'active_channel_selection': 'on',\n",
       " 'base_calling': 'on',\n",
       " 'reference_files': ['/data/ADAPTIVE_SAMPLING_iHSC/2000_closest.fasta'],\n",
       " 'filter_type': 'enrich',\n",
       " 'first_channel': '1',\n",
       " 'last_channel': '512',\n",
       " 'fast5_reads_per_file': '4000',\n",
       " 'fastq_reads_per_file': '4000',\n",
       " 'events': [[1, 512]],\n",
       " 'read_table': [[1, 512]],\n",
       " 'mux_scan_period': '1.5',\n",
       " 'pore_reserve': 'on',\n",
       " 'guppy_filename': 'dna_r9.4.1_450bps_hac.cfg',\n",
       " 'min_read_length': '200'}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parse_config(t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bc318da-1f2c-4042-954d-b17139f8667f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
